shared/src/metabase/shared/util/parameters.cljc

```diff
@@ -0,0 +1,145 @@
+             (if (.isValid m) (.format m "[Q]Q, YYYY") ""))
+     :clj value))
+
+(defmethod formatted-value :date/range
+  [_ value]
+  (let [[start end] (str/split value "~")]
+    (if (and start end)
+      (str (formatted-value :date/single start)
+           " - "
+           (formatted-value :date/single end))
+      "")))
+
+(defmethod formatted-value :date/relative
+  [_ value]
+  (case value
+    "today"      (trs "Today")
+    "yesterday"  (trs "Yesterday")
+    "past7days"  (trs "Past 7 Days")
+    "past30days" (trs "Past 30 Days")
+    "lastweek"   (trs "Last Week")
+    "lastmonth"  (trs "Last Month")
+    "lastyear"   (trs "Last Year")
+    "thisday"    (trs "Today")
+    "thisweek"   (trs "This Week")
+    "thismonth"  (trs "This Month")
+    "thisyear"   (trs "This Year")
+    ;; Always fallback to default formatting, just in case
+    (formatted-value :default value)))
+
+(defmethod formatted-value :date/all-options
+  [_ value]
+  ;; Test value against a series of regexes (similar to those in metabase/parameters/utils/mbql.js) to determine
+  ;; the appropriate formatting, since it is not encoded in the parameter type.
+  ;; TODO: this is a partial implementation that only handles simple dates
+  (condp (fn [re value] (->> (re-find re value) second)) value
+    #"^~?([0-9-T:]+)~?$"       :>> (partial formatted-value :date/single)
+    #"^([0-9-T:]+~[0-9-T:]+)$" :>> (partial formatted-value :date/range)
+    (str value)))
+
+(defmethod formatted-value :default
+  [_ value]
+  (cond
+    (and (sequential? value) (> (count value) 1))
+    (formatted-list value)
+
+    (sequential? value)
+    (str (first value))
+
+    :else
+    (str value)))
+
+(def ^:private escaped-chars-regex
+  #"[\\/*_`'\[\](){}<>#+-.!$@%^&=|\?~]")
+
+(defn- escape-chars
+  [text]
+  (str/replace text escaped-chars-regex #(str \\ %)))
+
+(defn- replacement
+  [tag->param match]
+  (let [tag-name (second match)
+        param    (get tag->param tag-name)
+        value    (:value param)
+        tyype    (:type param)]
+    (if value
+      (-> (formatted-value tyype value)
+          escape-chars)
+      ;; If this parameter has no value, return the original {{tag}} so that no substitution is done.
+      (first match))))
+
+(defn- normalize-parameter
+  "Normalize a single parameter by calling [[mbql.normalize/normalize-fragment]] on it, and converting all string keys
+  to keywords."
+  [parameter]
+  (->> (mbql.normalize/normalize-fragment [:parameters] [parameter])
+       first
+       (reduce-kv (fn [acc k v] (assoc acc (keyword k) v)) {})))
+
+(def ^:private template-tag-regex
+  "A regex to find template tags in a text card on a dashboard. This should mirror the regex used to find template
+  tags in native queries, with the exception of snippets and card ID references (see the metabase-lib function
+  `recognizeTemplateTags` for that regex)."
+  #"\{\{\s*([A-Za-z0-9_\.]+?)\s*\}\}")
+
+(defn ^:export substitute_tags
```

Comment:
super nit: I feel like this is the most important function of this namespace, so maybe move it to the bottom of the file?